whylog/constraints/verifier.py

```diff
@@ -144,37 +156,72 @@ def constraints_or(cls, clues_lists, effect, constraints, constraint_manager):
         and for each such combination produces InvestigationResult object.
         returns list of all produced InvestigationResults
         """
+        causes = []
+        if not constraints:
+            # when there is lack of constraints, but there are existing clues combinations,
+            # each of them should be returned
+            for combination in cls._clues_combinations(clues_lists):
+                causes.append(cls._pack_results_for_constraint_or(combination, constraints))
+        else:
+            clues_lists = cls._construct_proper_clues_lists(clues_lists)
+            for combination in cls._clues_combinations(clues_lists):
+                verified_constraints = [
+                    constraint
+                    for constraint in constraints
+                    if cls._verify_constraint(combination, effect, constraint, constraint_manager)
+                ]
+                if len(verified_constraints) > 0:
+                    causes.append(
+                        cls._pack_results_for_constraint_or(
+                            combination, verified_constraints
+                        )
+                    )
+        return causes
+
+    @classmethod
+    def single_constraint_not(cls, clues_lists, effect, constraint, constraint_manager):
```

Comment:
Good point. I assume that NOT has no sense when there are more than 1 constraint. But has sense when there is 0 constraints - but this function works for given constraint, so someone who will call this function, will have to previously check if there is any constraint and if so, then call this function. This functionality will be probably implemented soon when rule will be improved for handling OR and NOT (now it assumes that all constraints are linkaged with AND).